File: docs/_semantic_template_reference.js
/**
 * Module brief.
 * #region moduleContract
 * @modulecontract
 * @purpose [Describe the GOAL of the module — what business/operational need it fulfills, why.]
 * @scope
 *  - [Main functional areas covered by the module.]
 *  - NOT: [What is out of scope]
 * @invariants
 *  - [Condition/State that always holds]
 * @usecases
 *  - [Entity]: [Actor] => [Action] => [Goal]
 * @dependencies [Non-trivial deps - USES API: ..., READS: ..., WRITES: ...]
 * @rationale
 *  - Q: [Why was it implemented this way?]
 *    A: [Justification, environmental constraints.]
 * @keywords [Comma-separated keywords for grep search]
 * #endregion moduleContract
 **/

// this is just for decorator example
function frozen(target) {
  Object.freeze(target);
  Object.freeze(target.prototype);
}

// #region CLASS_ExampleClass
/**
 * Class brief.
 *
 * @purpose [Goal of the class and why — what it enables the user/agent to do.]
 */
@frozen
export class ExampleClass {
  privateMethodMarkupNotNeededForTrivialMethods() {
    return 1;
  }

  // #region METHOD_exampleMethod
  /**
   * Optional brief.
   *
   * @purpose [Goal of the method, why]
   * @requires [Optional preconditions]
   * @ensures [Optional postconditions]
   * @rationale [Optional rationale]
   * @param {number} param1 - Optional purpose/semantic meaning of input if non-trivial.
   * @returns {string} Optional purpose/semantic meaning of output if non-trivial.
   */
  exampleMethod(param1) {
    // #region BLOCK_calculateRegression Non trivial big block summary
    this.privateMethodMarkupNotNeededForTrivialMethods();
    // ... skip 10 lines
    console.debug("Pre calculation", { event: "start", param1 }); // block name + structured data
    // ... skip 10 lines
    const result = String(param1 * 2);
    // #endregion BLOCK_calculateRegression
    console.debug("Business result", { result });
    return result;
  }
  // #endregion METHOD_exampleMethod
}
// #endregion CLASS_ExampleClass

function privateFunctionMarkupNotNeededForTrivialFunctions() {
  // no-op
}

// #region FUNC_exampleFunction
/**
 * @purpose [Describe the GOAL of this function and why.]
 */
export function exampleFunction(data) {
  console.debug("INIT", { dataLength: data.length });
  privateFunctionMarkupNotNeededForTrivialFunctions();

  // #region BLOCK_loop
  let total = 0;
  for (const item of data) {
    total += item;
    console.debug("Loop step", { item, runningTotal: total });
  }
  // #endregion BLOCK_loop

  console.debug("RESULT", { sumComputed: total });
  return total;
}
// #endregion FUNC_exampleFunction

File: docs/_semantic_template_reference.cpp
/**
 * Module brief.
 * #region MODULE_CONTRACT
 * PURPOSE: [Describe the GOAL of this implementation — what business/operational need it fulfills, why.]
 * SCOPE:
 *   - [Main implementation responsibilities.]
 *   - NOT: [Public interface declarations or responsibilities owned by another module.]
 * INVARIANTS:
 *   - [Condition/state preserved while this module runs.]
 * USECASES:
 *   - [Entity]: [Actor] => [Action] => [Goal]
 * DEPENDENCIES: [Non-trivial deps — USES API: ..., READS: ..., WRITES: ...]
 * RATIONALE:
 *   - Q: [Why is this implementation organized this way?]
 *     A: [Justification and environmental constraints.]
 * KEYWORDS: [Comma-separated keywords for grep search]
 * #endregion MODULE_CONTRACT
 */

#include "example_class.h"

namespace {

void privateFunctionMarkupNotNeededForTrivialFunctions() {
  // no-op
}

}  // namespace

// #region METHOD_exampleMethod
/**
 * Optional brief.
 *
 * PURPOSE: [Goal of the method and why.]
 * REQUIRES: [Optional preconditions.]
 * ENSURES: [Optional postconditions.]
 * RATIONALE: [Optional rationale.]
 * @param param1 [Optional semantic meaning of a non-trivial input.]
 * @return [Optional semantic meaning of a non-trivial output.]
 */
String ExampleClass::exampleMethod(int param1) {
  // #region BLOCK_calculateResult Transform the validated input into the caller's result.
  privateMethodMarkupNotNeededForTrivialMethods();
  Serial.printf("event=example_calculation_start param1=%d\n", param1);
  const String result = String(param1 * 2);
  // #endregion BLOCK_calculateResult

  Serial.printf("event=example_calculation_result result_length=%u\n",
                static_cast<unsigned>(result.length()));
  return result;
}
// #endregion METHOD_exampleMethod

// #region FUNC_exampleFunction
/**
 * PURPOSE: [Describe the GOAL of this public function and why.]
 * REQUIRES: [Optional preconditions.]
 * ENSURES: [Optional postconditions.]
 * @param data [Optional semantic meaning of a non-trivial input.]
 * @return [Optional semantic meaning of a non-trivial output.]
 */
String exampleFunction(const String& data) {
  Serial.printf("event=example_function_start data_length=%u\n",
                static_cast<unsigned>(data.length()));
  privateFunctionMarkupNotNeededForTrivialFunctions();

  // #region BLOCK_buildResult Accumulate a result for each input character.
  size_t total = 0;
  for (size_t index = 0; index < data.length(); ++index) {
    ++total;
    Serial.printf("event=example_function_step index=%u running_total=%u\n",
                  static_cast<unsigned>(index), static_cast<unsigned>(total));
  }
  // #endregion BLOCK_buildResult

  Serial.printf("event=example_function_result sum=%u\n",
                static_cast<unsigned>(total));
  return String(total);
}
// #endregion FUNC_exampleFunction
